src/instructions.rs

```diff
@@ -0,0 +1,166 @@
+    }
+
+    /// Digest data, updating the internal state.
+    pub fn update(
+        &mut self,
+        mut layouter: impl Layouter<F>,
+        mut data: &[Sha256Chip::BlockWord],
+    ) -> Result<(), Error> {
+        self.length += data.len() * 32;
+
+        // Fill the current block, if possible.
+        let remaining = BLOCK_SIZE - self.cur_block.len();
+        let (l, r) = data.split_at(min(remaining, data.len()));
+        self.cur_block.extend_from_slice(l);
+        data = r;
+
+        // If we still don't have a full block, we are done.
+        if self.cur_block.len() < BLOCK_SIZE {
+            return Ok(());
+        }
+
+        // Process the now-full current block.
+        self.state = self.chip.compress(
+            &mut layouter,
+            &self.state,
+            self.cur_block[..]
+                .try_into()
+                .expect("cur_block.len() == BLOCK_SIZE"),
+        )?;
+        self.cur_block.clear();
+
+        // Process any additional full blocks.
+        let mut chunks_iter = data.chunks_exact(BLOCK_SIZE);
+        for chunk in &mut chunks_iter {
+            self.state = self.chip.initialization(&mut layouter, &self.state)?;
+            self.state = self.chip.compress(
+                &mut layouter,
+                &self.state,
+                chunk.try_into().expect("chunk.len() == BLOCK_SIZE"),
+            )?;
+        }
+
+        // Cache the remaining partial block, if any.
+        let rem = chunks_iter.remainder();
+        self.cur_block.extend_from_slice(rem);
+
+        Ok(())
+    }
+
+    /// Retrieve result and consume hasher instance.
+    pub fn finalize(
+        mut self,
+        mut layouter: impl Layouter<F>,
+    ) -> Result<Sha256Digest<Sha256Chip::BlockWord>, Error> {
+        // Pad the remaining block
+        if !self.cur_block.is_empty() {
+            let padding = vec![Sha256Chip::BlockWord::default(); BLOCK_SIZE - self.cur_block.len()];
+            self.cur_block.extend_from_slice(&padding);
+            self.state = self.chip.initialization(&mut layouter, &self.state)?;
+            self.state = self.chip.compress(
+                &mut layouter,
+                &self.state,
+                self.cur_block[..]
+                    .try_into()
+                    .expect("cur_block.len() == BLOCK_SIZE"),
+            )?;
+        }
+        self.chip
+            .digest(&mut layouter, &self.state)
+            .map(Sha256Digest)
+    }
+
+    /// Convenience function to compute hash of the data. It will handle hasher creation,
+    /// data feeding and finalization.
+    pub fn digest(
+        chip: Sha256Chip,
+        mut layouter: impl Layouter<F>,
+        data: &[Sha256Chip::BlockWord],
+    ) -> Result<Sha256Digest<Sha256Chip::BlockWord>, Error> {
+        let mut hasher = Self::new(chip, layouter.namespace(|| "init"))?;
+        hasher.update(layouter.namespace(|| "update"), data)?;
+        hasher.finalize(layouter.namespace(|| "finalize"))
+    }
```

Comment:
These are the instructions for SHA256, during our next meeting we can aim to build out the logic for the BLAKE2 instructions

Reply:
yes, so was trying to remove just the compression for blake2f, the issue is that it references functions in the instructions, its taking some time to see what I can safely remove out

Reply:
I see! yeah we'll likely have to create our own object structure for it to work and just follow the logic